### PR TITLE
Introduce new statement: SWITCH

### LINKED ISSUE #186: Add Switch&Case
We have IF statement rn, but It's not enough to handle  conditioning all situations. I think It's better to add new operator, SWITCH, CASE, which do the  same thing as switch(){ case: } in Java.

It could be LongRun, or might not be.

### LINKED ISSUE #495: Proposals: Add "SWITCH", "MATCH" statements
# NOTE: This is a draft issue.

### 💡 Motivation

Since TriggerReactor does not have statements similar to `switch` and `match`, I would like to suggest these.
Of course, you can even use `if-else` to challenge dealing with `switch` and/or `match` too, but I think it'd better to implement this for readability.

### 📜 Example

**Switch**:
```
IMPORT org.bukkit.GameMode

SWITCH player.getGameMode()
  CASE GameMode.CREATIVE
    #MESSAGE "You are in creative mode!"
    #BREAK
  CASE GameMode.SURVIVAL
    #MESSAGE "You are in survival mode!"
    #BREAK
  DEFAULT
    #MESSAGE "You are not in creative or survival mode!"
    #BREAK
ENDSWITCH
```

**Match**:
```
IMPORT org.bukkit.GameMode

playerGameModeInNumeric = MATCH player.getGameMode()
  CASE GameMode.CREATIVE -> 1
  CASE GameMode.SURVIVAL -> 0
  DEFAULT -> -1
ENDMATCH
```

### ❓ Why switch and match?

*TODO*

### ➕ What kinds of reserved words we should attention?

**[ Shared reserved words ]**
* **CASE**: Represent branch which satisfying the conditions.
* **DEFAULT**: Represent default branch which not satisfied with any conditions.

**[ Switch reserved words ]**
* **SWITCH**: Startpoint of switch-case.
* **ENDSWITCH**: Endpoint of switch-case. All possible values should be covered, or be provide `DEFAULT` case if any no matching arm is evaluated.

**[ Match reserved words ]**
* **MATCH**: Startpoint of match statement.
* **ENDMATCH**: Endpoint of match statement. All possible values should be covered, or be provide `DEFAULT` case if any no matching arm is evaluated.

**[ Match operators ]**
* **->**: A special operator `match` statement only for. Due to it should be designed as simple as possible, and `LAMBDA(=>)` statement only consumes factor-like tokens.

### PR DESCRIPTION
**Pull request by Sayakie**
This pull request adds a switch statement. Closes #186 and also closes #495.

## 🌟 Syntaxes

* Add new reversed keywords: `SWITCH`, `CASE`, `ENDCASE`, `DEFAULT`.
* Switch statement must start with `SWITCH`.
* Switch statement must end with `ENDSWITCH`.
* Each case clause can be ended with `ENDCASE` (optional).
* Use `CASE` for matching the expression's value to a `CASE` clause, it executes _statements_ associated with that case as well as the first arm is evaluated.
* No need to specify all possible values to be covered, but throw a `No matched arm` exception.
* Can use RANGE token for matching the expressions' value in case body.
* `DEFAULT` token stands for a `default` clause to be executed if the value of `expression` does not match any of the `case` clauses.
* Can use underscore(`_`) instead of `DEFAULT` token.
* Multiple match patterns may be joined with the `,` operator. Each pattern will be tested in a left-to-right sequence until a successful match is found.

## 📝 Examples

You can use switch statement in TriggerReactor like the following codes:

```js
IMPORT org.bukkit.GameMode

SWITCH player.getGameMode()
  CASE GameMode.SURVIVAL => #MESSAGE "You are in survival!"
  CASE "CREATIVE" =>
    #MESSAGE "You are in creative mode! Get some diamonds!"
    #GIVE "DIAMOND", random(20, 33)
  ENDCASE
  CASE ADVENTURE => #MESSAGE "You are an adventure! Good luck."
  DEFAULT => #MESSAGE "I guess you are a spectator... Maybe"
ENDSWITCH
```

```js
SWITCH random(0, 21)
  CASE 0..10 => #MESSAGE "Hello, you are between 0 and 9!"
  CASE 10..=17 =>
    #MESSAGE "Lucky guy! Get some diamonds!"
    #GIVE "DIAMOND", 8
  ENDCASE
  CASE 18, 19 => #MESSAGE "Hi, you chose 18 or 19."
  _ => #MESSAGE "Assume you chose 20. Probably..."
ENDSWITCH
```

```js
dice = random(1, 7)

SWITCH dice
  CASE 1, 2, 3 =>
    #MESSAGE "꽝! 다음 기회에..."
    #STOP
  CASE 4, 5 =>
    amount = random(16, 33)
    #MESSAGE "당첨! " + amount + "개의 다이아몬드를 얻었어요."
    #GIVE "DIAMOND", amount
ENDSWITCH
```
^ Note that if the dice is `6` the interpreter raises a `No matched arm` exception.

### CODE REVIEW

File: core/src/main/java/io/github/wysohn/triggerreactor/core/script/interpreter/Interpreter.java [hunk 1]
```diff
@@ -406,6 +406,185 @@ private void start(Node node, InterpreterLocalContext localContext) throws Inter
...
+                            start(caseBody, localContext);
+                            break iterCaseOrDefaultNodes;
+                        }
+                    }
+                }
+            }
+
+            if (!matches) {
+                // TODO(Sayakie): Raises an exception or ignore silently?
+                throw new InterpreterException("No matched arm");
```
Comment (wysohn) [inline core/src/main/java/io/github/wysohn/triggerreactor/core/script/interpreter/Interpreter.java:586]: I think it's okay to throw an exception here, given that we already make sure that there is at least one arm exist for the SWITCH in the parsing step

**Review by wysohn — COMMENTED** [review]
Looking good! Thank you a lot

Now the question is, should we push it to 3.4.x or 3.5.x?

I think we should focus on stabilizing 3.4.x than adding new features, and we can push this to 3.5.x

But let me know if you think otherwise

**Comment by Sayakie** [comment]
> Now the question is, should we push it to 3.4.x or 3.5.x?
> 
> I think we should focus on stabilizing 3.4.x than adding new features, and we can push this to 3.5.x
> 
> But let me know if you think otherwise

I think it would be good to bump with the 3.4 version.

We are currently facing some problems with core and managers, but not with parser. And 3.4 is not officially published and under the beta.

**Review by wysohn — APPROVED** [review]
Sounds good, lets go for it!

**Pull request merged**